x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -311,6 +333,49 @@ private void getVersionedDocument(String uid, ActionListener<VersionedDocument>
         });
     }
 
+    private void getVersionedDocuments(Collection<String> uids, ActionListener<Tuple<List<ProfileDocument>, Set<String>>> listener) {
+        if (uids.isEmpty()) {
+            listener.onResponse(new Tuple<>(List.of(), Set.of()));
+            return;
+        }
+        tryFreezeAndCheckIndex(listener).ifPresent(frozenProfileIndex -> {
+            frozenProfileIndex.checkIndexVersionThenExecute(
+                listener::onFailure,
+                () -> new OriginSettingClient(client, getActionOrigin()).prepareMultiGet()
+                    .addIds(frozenProfileIndex.aliasName(), uids.stream().map(ProfileService::uidToDocId).toArray(String[]::new))
+                    .execute(ActionListener.wrap(multiGetResponse -> {
+                        List<ProfileDocument> retrievedDocs = new ArrayList<>(multiGetResponse.getResponses().length);
+                        // ordered for tests
+                        Set<String> failures = new TreeSet<>();
+                        Exception loggedException = null;
+                        for (MultiGetItemResponse itemResponse : multiGetResponse.getResponses()) {
+                            if (itemResponse.isFailed()) {
+                                failures.add(docIdToUid(itemResponse.getId()));
+                                if (logger.isDebugEnabled() && itemResponse.getFailure().getFailure() != null) {
+                                    loggedException = ExceptionsHelper.useOrSuppress(
+                                        loggedException,
+                                        itemResponse.getFailure().getFailure()
+                                    );
+                                }
+                            } else if (itemResponse.getResponse() != null) {
+                                if (itemResponse.getResponse().isExists()) {
+                                    retrievedDocs.add(buildProfileDocument(itemResponse.getResponse().getSourceAsBytesRef()));
+                                } else if (logger.isDebugEnabled()) {
+                                    logger.debug("Profile [{}] not found", docIdToUid(itemResponse.getId()));
+                                }
+                            } else {
+                                logger.error("Inconsistent get item response [{}] [{}]", itemResponse.getIndex(), itemResponse.getId());
```

Comment:
I think we should add an assertion here so it fails loudly in CI if ever happens.

Reply:
Pushed [98f9ff4](https://github.com/elastic/elasticsearch/pull/85898/commits/98f9ff437f5950bc69a636a246e16a091a350e38)